Addition of SOP documentation (#126)
RedHatInsights/clowder: docs/sop.rst

Operating Clowder
=================

Two primary aspects of operating Clowder: Operating the apps managed by Clowder and operating
Clowder itself.

> Contrary to many other deployment models where apps configure themselves, an application that is

Reviewer: "configure themselves" sounds odd.  I wonder if saying something more like "where devs are burdened with the repsponsibility of wiring up all aspects of their applications in k8s", though that is much more wordy.

Author: Agreed

> managed by Clowder will have configuration presented to it and is expected to use it. Clowder

Reviewer: apps are "expected" to use it because there is no other way to obtain this information.  I know this is an SOP and not marketing material, but I would like to use this paragraph to emphasize the core value of Clowder by explicitly stating how we are removing the configuration effort from devs.

governs many different aspects of an application's configuration, from defining the port it should
listen on, to metrics, kafka topics, service hostnames, and more. In this way a common configuration
format is presented to the application no matter the envrionment it is running in, enabling a far

> easier development experience.

Reviewer: ok yeah you get to the benefits here, but I wonder if this should be explained at the beginning of the paragraph rather than the end?

Operating Apps Managed by Clowder
---------------------------------

ClowdEnvironment
++++++++++++++++

> **abbreviated to ``env`` in k8s**

Reviewer: ugh I wish the back ticks worked in bold text, but alas...

Author: I know it's sad :(

> The ``ClowdEnvironment`` CRD is responsible for configuring the environment that the Clowder enabled

Reviewer: what is an environment comprised of?  From Clowder's perspective, an environment is made up of the configuration and deployments needed to operate a complete clouddot environment.  I guess just saying "environment" here is kind of redundant.

apps will interact with. It is a *cluster scoped* CRD and thus must have a unique name inside the
k8s cluster. For production environments it is usual to have only one ``ClowdEnvironment``, whereas

> in other scenarios &mdash; such as ephemeral testing &mdash; Clowder enables the management of

Reviewer: I don't know how to make em-dashes in RST

multiple environments that operate completely independently from each other.

Providers
^^^^^^^^^

> Clowder provides a number of providers which govern the creation of services, e.g. Kafka topics,

Reviewer: ```suggestion
An environment's specification is broken into **providers**, which govern the creation of services, e.g. Kafka topics,
```

object storage, etc, that applications may depend on. The ``ClowdEnvironment`` CRD configures these

> providers principally by making use of a provider's mode.

Reviewer: ```suggestion
providers principally by making use of a provider's **mode**.
```

Modes
^^^^^

> Providers often operate in different modes. As an example the Kafka provider can operate in three

Reviewer: This first sentence seems redundant with the last sentence of the previous paragraph.  Maybe remove that last sentence??

Author: Sure - but if a section is navigated to having a brief context isn't a bad idea.

List of providers:

- `Kafka <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-databaseconfig>`_
- `Logging <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-loggingconfig>`_
- `Object storage <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-objectstoreconfig>`_
- `In-memory DB <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-inmemorydbconfig>`_
- `Relational DB <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-databaseconfig>`_
- `Web <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-webconfig>`_
- `Metrics <https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-metricsconfig>`_

Target Namespace
^^^^^^^^^^^^^^^^

> Clowder will provision two types of resources during operation, *environmental* and *application*

Reviewer: I think we can omit the app-specific info here since we're still under the ClowdEnv section.

specific. Application resources will be deployed into the same namespace the ClowdApp resides in.
Environmental resources, such as the Kafka/Zookeeper from the exmaple in the *Modes* section, will
be placed in the ``ClowdEnvironment``'s target namespace. This is configured by setting the

> ``targetNamespace`` attribute of the ``ClowdEnvironment`` and should always be set for production

Reviewer: maybe remove the assertion that it should be set in prod? Seems unnecessary.

environments. If it is omitted, a random target namespace is generated instead. The name of this
resource can be found by inspecting the ``status.targetNamespace`` of the ClowdEnvironment resource.

ClowdApp
++++++++

**abbreviated to ``app`` in k8s**

The ``ClowdApp`` CRD is responsible for configuring an application and is namespace scoped. Any
resources Clowder creates on behalf of the application will reside in the same namespace that the
``ClowdApp`` resources is applied to. As such the ``ClowdApp`` name must be unique within a
particular namespace.  An ``ClowdApp`` does not have to be placed in the ``ClowdEnvironment``'s
target namespace.

> A ``ClowdApp`` may define multiple services inside it. These services, though defined by a

Reviewer: You refer to the `pod` specs as "services", causing an inconsistency.  Is that what we want to call them?

Author: I'm not sure, should we do a spec change to bring the terminology inline before this is used outside of the platform?

Reviewer: Yeah but we should make sure we choose a good name, which is a challenge.

Author: Agreed......

specification that is very similar to the k8s pod specification, will be deployed as individual
deployment resources.  Functionally, defining multiple applications in the same ``ClowdApp``
specification allows the sharing of some infrastructure dependencies such as databases.

> Applications in different ClowdApp's should not expect to be able to share databases.

Reviewer: hmm I don't know if this is true anymore.  I think it may be hard to hold this assertion as more apps onboard

Author: What do you mean?

``ClowdApp`` spec. More information on each of them is defined in the `API specification`_.

.. _API specification: https://redhatinsights.github.io/clowder/api_reference.html#k8s-api-cloud-redhat-com-clowder-v2-apis-cloud-redhat-com-v1alpha1-clowdappspec

Created Resources
^^^^^^^^^^^^^^^^^

For each ``ClowdApp`` service, Clowder will create an ``apps.Deployment`` and a ``Service``
resource.  If the service has the ``web`` field set to true, the ``Service`` resource will
include a port definition for ``webPort`` as well as the standard ``metricsPort``. The actual values
of these are defined in the ``ClowdEnvironment`` by configuring the web and metric providers,
respectively. By default these are set to 8000 for the web service port and 9000 for the metrics
port.

Clowder will also set certain fields in the pod spec, inline with best practice, such as pull
policy, and anti-affinity.

Clowder will also create a ``Secret`` resource which will contain the generated configuration
for that app. This secret will be mounted at ``/cdappconfig.json`` and will be consumed by the app
to configure itself on startup.

> 

Reviewer: What about local redis and postgresql DBs?
